syntax = "proto3";

package bacs.archive.problem;

import "bacs/problem/problem.proto";

message IdSet {
  repeated string id = 1;
}

message Info {
  oneof result {
    string error = 1;
    bacs.problem.Problem problem = 2;
  }
}

message InfoMap {
  map<string, Info> entry = 1;
}

message FlagSet {
  repeated Flag flag = 1;
}

// It is OK to specify reserved flag as custom.
message Flag {
  enum Reserved {
    // Reserved default value, should not be used.
    NONE = 0;
    // This flag will exclude problem from non-flags and non-status functions.
    // It is set automatically if problem can't be repacked.
    IGNORE = 1;
    // Exclude problem from operations that
    // can alter problem's archive or problem id, i.e. user
    // will not be able to re-upload problem, rename or remove it.
    // But it is still possible to change problem's flags and repack it.
    LOCKED = 2;
    // Harder version of LOCKED.
    // It does not allow to edit flags.
    // Only administrator can remove this flag manually.
    READ_ONLY = 3;
    // Repack is planned and will be done soon.
    PENDING_REPACK = 4;
  }
  oneof flag {
    Reserved reserved = 1;
    string custom = 2;
  }
}

message Status {
  FlagSet flags = 1;
  bacs.problem.Revision revision = 2;
}

message ImportInfo {
  oneof result {
    string error = 1;
    Status status = 2;
  }
}

message ImportMap {
  map<string, ImportInfo> entry = 1;
}
